### PR TITLE
Add test integration

### PR DESCRIPTION
**Pull request by futuretea**
#45

* Refactor some setup register entrances to support configuring in integration tests.
* Introduce Ginkgo and Kind as the testing base, and make an integration test framework for the Ginkgo suite entrance.
* Use Ginkgo syntax to write some cases.

How to run:
  * Configure the appropriate environment variables
    - If there is an existing  cluster, we can locate the KubeConfig to local and `export USE_EXISTING_CLUSTER=true` to use it, default is to create a  local Kind cluster. (Tips) In local test, you can `export KIND_IMAGE_MIRROR="your image mirror url" `to speed up .
    - If there is an existing harvester runtime in the cluster, we can `export USE_EXISTING_RUNTIME=true` to use it, default is to install a harvester runtime to cluster or update the harvester runtime in cluster.
    - If the harvester runtime suuport kvm, we can `export DONT_USE_EMULATION=true` to support it, default is to use qemu software emulation.
    - If we want to review testing cluster, we can `export KEEP_TESTING_CLUSTER=true`  to keep it , default is to delete the local Kind cluster after test finished.
    - If we want to review testing runtime, we can `export KEEP_TESTING_RUNTIME=true`  to keep it, default is to delete the harvester runtime after test finished.
    - If we want to review testing vm, we can `export KEEP_TESTING_VM=true` to keep it, default is to delete each testing vm after each test case finished.
  * Run ./script/test-integration or make test-integration.

### CODE REVIEW

**Review by gitlawr — COMMENTED** [review]
Can we add a readme file with the info provided in the PR description? So that people know how to run the tests.

**Comment by gitlawr** [comment]
And move from `test` to `tests` directory

**Comment by futuretea** [comment]
> Can we add a readme file with the info provided in the PR description? So that people know how to run the tests.

done

File: Dockerfile.dapper [hunk 1]
```diff
@@ -28,9 +28,14 @@ RUN if [ "$(go env GOARCH)" = "amd64" ]; then \
...
         GO111MODULE=on go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.3.0; \
     fi
+## install ginkgo
+RUN if [ "$(go env GOARCH)" = "amd64" ]; then \
+        GO111MODULE=on go get github.com/onsi/ginkgo/ginkgo@v1.14.0; \
+    fi
 # -- for make rules
 
 # -- for dapper
+ENV DAPPER_RUN_ARGS --privileged --network host
```
Comment (gitlawr) [inline Dockerfile.dapper:38]: Why do we need the host network? Is it dependent on the network setup? I don't see any prerequisite mentioned though

Comment (futuretea) [reply Dockerfile.dapper:38]: in the dapper container，harvester progress can be able to access the kind cluster  by host network.

Comment (futuretea) [reply Dockerfile.dapper:38]: image

**Review by gitlawr — APPROVED** [review]
LGTM

**Review by guangbochen — COMMENTED** [review]
additionally can u please help to change the tab default to 4 space thanks?

File: tests/integration/runtime/runtime.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+var (
+	testDeploymentManifest = []string{
+		"cdi-operator",
+		"cdi-apiserver",
+		"cdi-deployment",
+		"cdi-uploadproxy",
+		"virt-operator",
+		"virt-api",
+		"virt-controller",
+		"minio",
```
[the remark this answers is not in the document]
Comment (futuretea) [reply tests/integration/runtime/runtime.go:35]: the harvester progress is running outside the kind cluster

File: tests/integration/runtime/construct.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+	return nil
+}
+
+// installHarvesterChart installs the basic components of harvester.
+func installHarvesterChart(ctx context.Context, kubeConfig *restclient.Config) error {
+	// chart values patches
+	patches := map[string]interface{}{
+		"replicas":           0,
+		"minio.service.type": "NodePort",
```
[the remark this answers is not in the document]
Comment (futuretea) [reply tests/integration/runtime/construct.go]: fixed

File: tests/integration/runtime/destruct.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rancher/harvester/tests/framework/env"
+
+	"github.com/rancher/harvester/tests/framework/ready"
+
+	"github.com/rancher/harvester/tests/framework/helm"
```
[the remark this answers is not in the document]
Comment (futuretea) [reply tests/integration/runtime/destruct.go]: fixed

**Review by guangbochen — APPROVED** [review]

**Pull request merged**